core/src/main/scala/kafka/log/LogLoader.scala

```diff
@@ -106,7 +158,14 @@ object LogLoader extends Logging {
       loadSegmentFiles(params)
     })
 
-    completeSwapOperations(swapFiles, params)
+    // Forth pass: rename remaining index swap files. They must be left due to a broker crash when
```

Comment:
Hmm, not sure why we still have swap files at the point. We have renamed all existing swap files and no new swap files are created.

Reply:
We have renamed all `.log.swap` files and their corresponding index swap files. If there is a single .index.swap file, it is not renamed previously in the recovery process. A single .index.swap file could happen if it crashed in the middle of this line: https://github.com/apache/kafka/blob/bd668e90c6992e7a1f23274aab37d186d6a4b560/core/src/main/scala/kafka/log/Log.scala#L2381

Reply:
In this line https://github.com/apache/kafka/pull/10763/files#diff-54b3df71b1e0697a211d23a9018a91aef773fca0b9cbd1abafbdca6c79664930R105 , we filtered out `.log.swap` files. Single `.index.swap` files (those without corresponding `.log.swap` files) are not in the set of `toRenameSwapFiles`, and therefore are not recovered in https://github.com/apache/kafka/pull/10763/files#diff-54b3df71b1e0697a211d23a9018a91aef773fca0b9cbd1abafbdca6c79664930R138

Reply:
Should we rename all `.swap` files in https://github.com/apache/kafka/pull/10763/files#diff-54b3df71b1e0697a211d23a9018a91aef773fca0b9cbd1abafbdca6c79664930R138 no matter the file is in `toRenameSwapFiles` or not? I am not sure if we have to put the `retryOnOffsetOverflow` call between renaming files in `toRenameSwapFiles` and renaming the rest `.swap` files.

Reply:
I guess we don't need to put `retryOnOffsetOverflow` call in between. Removed the `toRenameSwapFiles` variable and combined the renaming in https://github.com/apache/kafka/pull/10763/commits/76c197c87cbdd9a7e109ce6330a4e88eaf2ad911